### PR TITLE
Deployment of etcd and overall skeleton

### PR DESCRIPTION
**Pull request by achimweigel**
**How to categorize this PR?**

/area delivery
/kind enhancement

**What this PR does / why we need it**:

This pull request adds a skeleton for the overall deployment of virtual gardener with the landscaper. This includes the import and export interface of the landscaper component. Currently only the etcds and the service is deployed. Here only small modifications are done mainly the integration of changes from landscape setup.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

Most of the changed files are from the vendor folder

**Release note**:

```feature operator
NONE
```

### CODE REVIEW

File: .ci/component_descriptor [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+SOURCE_PATH="$(dirname $0)/.."
+COMMIT_SHA="$(git --git-dir ${SOURCE_PATH}/.git rev-parse HEAD)"
+
+if [[ ${RESOURCES_FILE_PATH} == "" ]]; then
+  RESOURCES_FILE_PATH="${SOURCE_PATH}/.landscaper/resources.yaml"
+fi
+
+printf "> Building Virtual Garden with version ${EFFECTIVE_VERSION} - ${COMMIT_SHA}\n"
+
+printf "> Create Virtual Garden ca archive\n"
```
Comment (rfranzke) [inline .ci/component_descriptor]: What is a "ca archive"?

File: .ci/verify [hunk 2]
```diff
@@ -6,11 +6,7 @@ set -o pipefail
 
 cd "$(dirname $0)/.."
 
-git config --global user.email "gardener@sap.com"
-git config --global user.name "Gardener CI/CD"
-
-mkdir -p /go/src/github.com/gardener/virtual-garden
-cp -r . /go/src/github.com/gardener/virtual-garden
-cd /go/src/github.com/gardener/virtual-garden
```
Comment (rfranzke) [inline .ci/verify]: Why was this changed?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# SPDX-FileCopyrightText: 2020 SAP SE or an SAP affiliate company and Gardener contributors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+#### BUILDER ####
+FROM golang:1.16 AS builder
```
Comment (rfranzke) [inline Dockerfile]: Can you specify a patch version?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+ARG EFFECTIVE_VERSION
+
+RUN make install EFFECTIVE_VERSION=$EFFECTIVE_VERSION
+
+#### BASE ####
+FROM eu.gcr.io/gardenlinux/gardenlinux:184.0 AS base
+
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get --yes -o Dpkg::Options::="--force-confnew" install ca-certificates \
+    && rm -rf /var/lib/apt /var/cache/apt
```
Comment (rfranzke) [inline Dockerfile]: Why is this needed?

File: example/imports.yaml [hunk 5]
```diff
@@ -1,32 +1,108 @@
...
 
 virtualGarden:
+  deleteNamespace: true
+
   etcd:
     storageClassName: my-virtual-garden-storage-class
     backup:
       infrastructureProvider: gcp # aws|gcp|alicloud
-      region: europe-west5
+      region: europe-west555
```
Comment (rfranzke) [inline example/imports.yaml]: ?

File: example/imports.yaml [hunk 6]
```diff
@@ -1,32 +1,108 @@
...
-  #     ttl: 120
+    replicas: 1
+    sni:
+      hostname: api.dev.gardener.cloud.sap
+      dnsClass: gardendns
+      ttl: 120
+      secretName: gardener-dashboard-tls
+    dnsAccessDomain: my.dns.domain
+
+    gardenerControlplane:
```
Comment (rfranzke) [inline example/imports.yaml:39]: What has the gardener control plane to do with the `kubeAPIServer` configuration?

File: pkg/api/exports.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+type Exports struct {
```
Comment (rfranzke) [inline pkg/api/exports.go:18]: doc string

File: pkg/api/hvpa_types.go [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package api
```
Comment (rfranzke) [inline pkg/api/hvpa_types.go:15]: license header (I don't repeat this for other files, so please check)

File: pkg/api/hvpa_types.go [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"github.com/gardener/hvpa-controller/api/v1alpha1"
+)
+
+type HvpaConfig struct {
```
Comment (rfranzke) [inline pkg/api/hvpa_types.go:22]: doc string for all exported types (I don't repeat this for other files, so please check)

File: pkg/api/loader/exports.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"sigs.k8s.io/yaml"
+
+	"github.com/gardener/virtual-garden/pkg/api"
+)
+
+func ExportsToFile(exports *api.Exports, path string) error {
```
Comment (rfranzke) [inline pkg/api/loader/exports.go:28]: Unit tests for these functions?

File: pkg/virtualgarden/utils.go [hunk 11]
```diff
@@ -17,6 +17,9 @@ package virtualgarden
 import (
 	"context"
 
+	"github.com/ghodss/yaml"
+	v1 "k8s.io/client-go/tools/clientcmd/api/v1"
```
Comment (rfranzke) [inline pkg/virtualgarden/utils.go:21]:
```suggestion
	"k8s.io/client-go/tools/clientcmd/api/v1"
```

File: .ci/component_descriptor (hunk 1 above)
Comment (achimweigel) [reply .ci/component_descriptor]: ca is a component archive (see https://gardener.github.io/component-spec/oci.html?highlight=archive). We change the text to `Create Virtual Garden Component Archive`.

File: .ci/verify (hunk 2 above)
Comment (achimweigel) [reply .ci/verify]: We reverted these changes

File: Dockerfile (hunk 3 above)
Comment (achimweigel) [reply Dockerfile]: We switched to 1.16.5.

File: Dockerfile (hunk 4 above)
Comment (achimweigel) [reply Dockerfile]: This was just a leftover of the origin. I removed this line.

File: example/imports.yaml (hunk 5 above)
Comment (achimweigel) [reply example/imports.yaml]: Just to be sure not to deploy something working. I have reverted to the origin.

File: pkg/api/exports.go (hunk 7 above)
Comment (achimweigel) [reply pkg/api/exports.go:18]: Doc string added.

File: pkg/api/hvpa_types.go (hunk 8 above)
Comment (achimweigel) [reply pkg/api/hvpa_types.go:15]: Licence header added to all files.

File: pkg/api/hvpa_types.go (hunk 9 above)
Comment (achimweigel) [reply pkg/api/hvpa_types.go:22]: Doc strings added for all in the api package.

File: pkg/virtualgarden/utils.go (hunk 11 above)
Comment (achimweigel) [reply pkg/virtualgarden/utils.go:21]: v1 is removed

File: pkg/api/loader/exports.go (hunk 10 above)
Comment (achimweigel) [reply pkg/api/loader/exports.go:28]: Some test added. The real functionality is tested in that part of the integration test, which is currently not included in this pull request.

**Review by rfranzke — APPROVED** [review]
